File: cmd_add.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/nektro/go-util/arrays/stringsu"
	"github.com/nektro/go-util/util"
	"github.com/spf13/cobra"
	"gopkg.in/yaml.v2"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{
		Use:   "add",
		Short: "Add a new dependency to your project",
		Long:  `zigmod add <type> <path>`,
		Run: func(cmd *cobra.Command, args []string) {
			assert(len(args) >= 1, "missing package <type> parameter")
			assert(len(args) >= 2, "missing package <path> parameter")

			dept := args[0]
			path := args[1]

			assert(stringsu.Contains(AllDepTypes, dept), "provided <type> parameter is not a valid dependency type")

			f, err := os.Open("./zig.mod")
			util.DieOnError(err)
			bys, _ := ioutil.ReadAll(f)

			m := new(ModFile)
			yaml.Unmarshal(bys, &m)

			for _, item := range m.Deps {
				if item.Type == dept && item.Path == path {
					assert(false, "dependency already added, skipping!")
				}
			}
			m.Deps = append(m.Deps, Dep{dept, path})

			f.Close()
			f, err = os.Create("./zig.mod")
			util.DieOnError(err)
			d, _ := yaml.Marshal(m)
			fmt.Fprint(f, string(d))
			log.Println("Successfully added", path)
		},
	})
}

File: src/cmd_add.zig
const std = @import("std");
const gpa = std.heap.c_allocator;

const u = @import("./util/index.zig");

//
//

pub fn execute(args: [][]u8) !void {
    //
    u.assert(args.len >= 1, "missing package <type> parameter", .{});
    u.assert(args.len >= 2, "missing package <path> parameter", .{});

    const dept = args[0];
    const path = args[1];

    const dep_type = std.meta.stringToEnum(u.DepType, dept);
    u.assert(dep_type != null, "provided <type> parameter \"{}\" is not a valid dependency type", .{dept});

    const m = try u.ModFile.init(gpa, "./zig.mod");
    for (m.deps) |d| {
        u.assert(!(d.type == dep_type.? and std.mem.eql(u8, d.path, path)), "dependency already added, skipping!", .{});
    }

    const ndl = &std.ArrayList(u.Dep).init(gpa);
    for (m.deps) |d| {
        try ndl.append(d);
    }
    try ndl.append(u.Dep{
        .type = dep_type.?,
        .path = path,
    });

    //
    const f = try std.fs.cwd().createFile("./zig.mod", .{});
    defer f.close();

    const w = f.writer();
    try w.print("name: {}\n", .{m.name});
    try w.print("main: {}\n", .{m.main});
    try w.print("dependencies:\n", .{});

    for (ndl.items) |d| {
        try w.print("- type: {}\n", .{@tagName(d.type)});
        try w.print("  path: {}\n", .{d.path});
    }

    u.print("Successfully added {}", .{path});
}
